contracts/root/TokenPredicates/ChainExitERC1155Predicate.sol

```diff
@@ -0,0 +1,168 @@
+        return ERC1155Receiver(0).onERC1155BatchReceived.selector;
+    }
+
+    /**
+     * @notice Lock ERC1155 tokens for deposit, callable only by manager
+     * @param depositor Address who wants to deposit tokens
+     * @param depositReceiver Address (address) who wants to receive tokens on child chain
+     * @param rootToken Token which gets deposited
+     * @param depositData ABI encoded id array and amount array
+     */
+    function lockTokens(
+        address depositor,
+        address depositReceiver,
+        address rootToken,
+        bytes calldata depositData
+    ) external override only(MANAGER_ROLE) {
+        // forcing batch deposit since supporting both single and batch deposit introduces too much complexity
+        (
+            uint256[] memory ids,
+            uint256[] memory amounts,
+            bytes memory data
+        ) = abi.decode(depositData, (uint256[], uint256[], bytes));
+
+        emit LockedBatchChainExitERC1155(
+            depositor,
+            depositReceiver,
+            rootToken,
+            ids,
+            amounts
+        );
+        IMintableERC1155(rootToken).safeBatchTransferFrom(
+            depositor,
+            address(this),
+            ids,
+            amounts,
+            data
+        );
+    }
+
+    function makeArrayWithValue(uint256 val, uint size) internal pure returns(uint256[] memory) {
+        require(
+            size > 0,
+            "ChainExitERC1155Predicate: Invalid resulting array length"
+        );
+
+        uint256[] memory vals = new uint256[](size);
+
+        for (uint256 i = 0; i < size; i++) {
+            vals[i] = val;
+        }
+
+        return vals;
+    }
+
+    /**
+     * @notice Validates log signature, withdrawer address
+     * then sends the correct tokenId, amount to withdrawer
+     * callable only by manager
+     * @param rootToken Token which gets withdrawn
+     * @param log Valid ChainExit log from child chain
+     */
+    function exitTokens(
+        address,
+        address rootToken,
+        bytes memory log
+    ) public override only(MANAGER_ROLE) {
+        RLPReader.RLPItem[] memory logRLPList = log.toRlpItem().toList();
+        RLPReader.RLPItem[] memory logTopicRLPList = logRLPList[1].toList();
+        bytes memory logData = logRLPList[2].toBytes();
+
+        if (bytes32(logTopicRLPList[0].toUint()) == CHAIN_EXIT_EVENT_SIG) {
+
+            address withdrawer = address(logTopicRLPList[1].toUint());
+            require(withdrawer != address(0), "ChainExitERC1155Predicate: INVALID_RECEIVER");
+
+            (uint256 id, uint256 amount, bytes memory data) = abi.decode(
+                logData,
+                (uint256, uint256, bytes)
+            );
+
+            IMintableERC1155 token = IMintableERC1155(rootToken);
+            uint256 tokenBalance = token.balanceOf(address(this), id);
+
+            if (tokenBalance < amount) {
+                token.mintBatch(
+                    address(this), 
```

Comment:
Would minting directly to the `withdrawer` here perhaps be an optimization? I don't think that there's a difference if only minting a single tokenId, but it may be more critical if minting a batch of different `tokenIds` & `amounts`?

Reply:
I'd like to check gas savings during execution. I'll implement that if I see improvement over current one.

Reply:
Hello, after testing both of following scenarios

- batch mint to self & finally transfer whole exitable amount to withdrawer 1️⃣
- batch mint extra to withdrawer & finally transfer only locked amount to withdrawer 2️⃣

I found 2️⃣ is efficient in terms of gas consumption.

---
Here's a snippet 👇

1️⃣

![image](https://user-images.githubusercontent.com/45074836/117407590-63a34b80-af2c-11eb-9fae-fa99ffe85589.png)

2️⃣

![image](https://user-images.githubusercontent.com/45074836/117407621-6ef67700-af2c-11eb-8eb2-8fd6d44a4f27.png)

---

I've planned to use 2️⃣ finally, commit to be pushed soon 😉.

Reply:
Changes pushed : https://github.com/maticnetwork/pos-portal/pull/77/commits/9dd095d75ea67bbfc15ddaf2cf44dec3af300dae

Reply:
@itzmeanjan looks goodThe extra `bytes data` being passed in both calls ( `mintBatch` & `safeBatchTransferFrom`) makes sense. We will definitely need the data when minting the new tokens.